File: ocaml/AC14.ml
let get grid transform_coords coords =
  let (i, j) = transform_coords coords in
  grid.(i).(j)

let set grid transform_coords coords value =
  let (i, j) = transform_coords coords in
  grid.(i).(j) <- value

let rec drop_rounded_rocks grid transform_coords column row count =
  if count > 0 then
    begin
      set grid transform_coords (row, column) 'O';
      drop_rounded_rocks grid transform_coords column (row + 1) (count - 1)
    end

let tilt_column grid transform_coords height column =
  let rounded_rock_count = ref 0 in
  for row = height - 1 downto 0 do
    match get grid transform_coords (row, column) with
    | 'O' ->
       incr rounded_rock_count;
       set grid transform_coords (row, column) '.'
    | '#' ->
       drop_rounded_rocks grid transform_coords column (row + 1)
         !rounded_rock_count;
       rounded_rock_count := 0
    | '.' -> ()
    | _ -> assert false
  done;
  drop_rounded_rocks grid transform_coords column 0 !rounded_rock_count

let tilt grid transform_coords height width =
  for column = 0 to width - 1 do
    tilt_column grid transform_coords height column
  done

let tilt_north grid =
  let height = Array.length grid in
  let width = Array.length grid.(0) in
  tilt grid Fun.id height width

let tilt_cycle grid =
  let height = Array.length grid in
  let width = Array.length grid.(0) in
  tilt grid Fun.id height width;
  tilt grid (fun (i, j) -> (j, i)) width height;
  tilt grid (fun (i, j) -> (height - i - 1, j)) height width;
  tilt grid (fun (i, j) -> (j, width - i - 1)) width height

let parse_line line =
  Array.of_seq (String.to_seq line)

let amount_of_load grid =
  let total = ref 0 in
  let height = Array.length grid in
  let width = Array.length grid.(0) in
  for row = 0 to height - 1 do
    for column = 0 to width -1 do
      if grid.(row).(column) = 'O' then
        total := !total + (height - row)
    done
  done;
  !total

let copy_grid grid =
  Array.copy (Array.map Array.copy grid)

let measure_iterate_tilting_cycle grid =
  let table = Hashtbl.create 16 in
  let rec loop count =
    match Hashtbl.find_opt table grid with
    | Some index -> (index, count)
    | None ->
       Hashtbl.add table (copy_grid grid) count;
       tilt_cycle grid;
       loop (count + 1) in
  loop 0

let iterate_tilting_cycle grid count =
  let (initial, count') = measure_iterate_tilting_cycle grid in
  assert (count >= count');
  let left = (count - count') mod (count' - initial) in
  for _ = 1 to left do
    tilt_cycle grid;
  done

let () =
  let grid = Array.of_seq (Seq.map parse_line (Common.input_lines ())) in
  let result_part1 =
    let grid_part1 = copy_grid grid in
    tilt_north grid_part1;
    amount_of_load grid_part1 in
  Format.eprintf "Part 1: %d@." result_part1;
  let result_part2 =
    let grid_part2 = copy_grid grid in
    iterate_tilting_cycle grid_part2 1000000000;
    amount_of_load grid_part2 in
  Format.eprintf "Part 2: %d@." result_part2

File: rust/src/bin/ac14.rs
type Grid = Vec<Vec<char>>;

fn drop_rounded_rocks(
    grid: &mut Grid, column: usize, row: usize, count: usize,
    transform: &impl Fn(usize, usize) -> (usize, usize)
) {
    for i in row .. row + count {
        let (x, y) = transform(column, i);
        grid[y][x] = 'O';
    }
}

fn tilt_column(
    grid: &mut Grid, height: usize, column: usize,
    transform: &impl Fn(usize, usize) -> (usize, usize)
) {
    let mut rounded_rock_count = 0;
    for row in (0 .. height).rev() {
        let (x, y) = transform(column, row);
        match grid[y][x] {
            'O' => {
                rounded_rock_count += 1;
                grid[y][x] = '.'
            }
            '#' => {
                drop_rounded_rocks(
                    grid, column, row + 1, rounded_rock_count, transform
                );
                rounded_rock_count = 0
            }
            '.' => (),
            _ => panic!("tilt_column")
        }
    }
    drop_rounded_rocks(grid, column, 0, rounded_rock_count, transform);
}

fn tilt(
    grid: &mut Grid, height: usize, width: usize,
    transform: &impl Fn(usize, usize) -> (usize, usize)
) {
    for column in 0 .. width {
        tilt_column(grid, height, column, transform)
    }
}

fn tilt_north(grid: &mut Grid) {
    let height = grid.len();
    let width = grid[0].len();
    tilt(grid, height, width, &|i, j| (i, j));
}

fn tilt_cycle(grid: &mut Grid) {
    let height = grid.len();
    let width = grid[0].len();
    tilt(grid, height, width, &|i, j| (i, j));
    tilt(grid, width, height, &|i, j| (j, i));
    tilt(grid, height, width, &|i, j| (i, height - j - 1));
    tilt(grid, width, height, &|i, j| (width - j - 1, i));
}

fn amount_of_load(grid: &Grid) -> usize {
    grid.iter().enumerate().map(|(i, line)|
      line.iter().filter(|&&c| c == 'O').count() * (grid.len() - i)
    ).sum()
}

fn measure_cycle<'a>(
    arena: &'a typed_arena::Arena<Grid>,
    grid: Grid,
    map: &mut std::collections::HashMap<&'a Grid, usize>,
    seq: &mut Vec<&'a Grid>,
) -> (usize, usize) {
    let mut grid_ref = arena.alloc(grid);
    loop {
        match map.entry(grid_ref) {
            std::collections::hash_map::Entry::Occupied(occupied) => {
                return (*occupied.get(), seq.len());
            }
            std::collections::hash_map::Entry::Vacant(vacant) => {
                vacant.insert(seq.len());
            }
        }
        seq.push(grid_ref);
        let mut new_grid = grid_ref.clone();
        tilt_cycle(&mut new_grid);
        grid_ref = arena.alloc(new_grid);
    }
}

fn iterate_tilt(grid: Grid, count: usize) -> usize {
    let arena = typed_arena::Arena::new();
    let mut map = std::collections::HashMap::new();
    let mut seq = Vec::new();
    let (initial, cycle) = measure_cycle(&arena, grid, &mut map, &mut seq);
    let grid =
        if count < cycle {
            seq[count]
        }
        else {
            let left = (count - cycle) % (cycle - initial);
            seq[initial + left]
        };
    amount_of_load(grid)
}

fn main() {
    let lines = std::io::stdin().lines();
    let grid: Grid =
        lines.map(|line| line.unwrap().chars().collect()).collect();
    let mut grid_part1 = grid.clone();
    tilt_north(&mut grid_part1);
    let result_part1 = amount_of_load(&grid_part1);
    println!("Part 1: {result_part1}");
    let result_part2 = iterate_tilt(grid, 1000000000);
    println!("Part 2: {result_part2}");
}
